python/dgl/backend/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,160 @@
+import torch as th
+from .tensor import attach_grad, is_recording
+from .util import get_shared_mem_array, create_shared_mem_array
+from datetime import timedelta
+
+from ... import ndarray as nd
+
+_store = None
+
+class NodeEmbedding: # NodeEmbedding
+    '''Sparse node embeddings for graph.
+
+    DGL provides a sparse embedding to support models that require learnable embeddings.
+    DGL's sparse embeddings are mainly used for learning node embeddings of graph models.
+    The sparse embeddings have to be updated by DGL's optimizers instead of
+    the optimizers provided by the deep learning frameworks (e.g., Pytorch).
+
+    To support efficient training on a graph with many nodes, the embeddings support sparse
+    updates. That is, only the embeddings involved in a mini-batch computation are updated.
+    Currently, DGL provides two optimizer: `SparseAdagrad` and `SparseAdam. DGL will provide more
+    optimizers in the future.
+
+    Parameters
+    ----------
+    num_embeddings : int
+        The number of embeddings. Currently, the number of embeddings has to be the same as
+        the number of nodes or the number of edges.
+    embedding_dim : int
+        The dimension size of embeddings.
+    name : str
+        The name of the embeddings. The name should uniquely identify embeddings in a system.
+    init_func : callable, optional
+        The function to create the initial data. If the init function is not provided,
+        the values of the embeddings are initialized to zero.
+    group : str, optional
+        Group name.
+    host_name : str, optional
+        The hostname or IP Address the server store should run on.
+    port : int, optional
+        The port on which the server store should listen for incoming requests.
+
+    Examples
+    --------
+    Before launching multiple gpu processes
+
+    >>> def initializer(emb):
+            th.nn.init.xavier_uniform_(emb)
+            return emb
+
+    In each multiple gpu process
+
+    >>> emb = dgl.GraphSparseEmbedding(g.number_of_nodes(), 10, 'emb', init_func=initializer)
+    >>> optimizer = dgl.SparseAdagrad([emb], lr=0.001)
+    >>> for blocks in dataloader:
+    ...     feats = emb(nids, gpu_0)
+    ...     loss = F.sum(feats + 1, 0)
+    ...     loss.backward()
+    ...     optimizer.step()
+    '''
+
+    def __init__(self, num_embeddings, embedding_dim, name,
+                 init_func=None, group=None, host_name='127.0.0.1', port=12346):
```

Comment:
if this is used in a single machine, why do we want users to change host_name or port?

Reply:
In case they are used by some other program.